### ISSUE TITLE
Unable to install dependencies for demo/node.js (Windows 10)  

### ISSUE THREAD

**Issue by jayprout** [vladmandic/face-api#88]
Hello there
My system is Windows 10 laptop (no GPU/Graphics card) 
I running node version 14.16.0

I'm trying to get a demo up and running of face-api running on node server side, not in an HTML browser. I can get the demo browser to work but not the demo/node.js version.

I entered a npm install command to install all the dependencies listed in package.json and I get the following error:
npm ERR! fatal: No url found for submodule path 'src' in .gitmodules

How do I fix this please?

Regards - Jay

**Comment by vladmandic** [comment]
**FaceAPI** does not use git *submodules* at all, so this is something local  

But you don't need all dependencies from `package.json`, only ones actually used by specific demo  

For example, `demo/node.js` requires:
- `@vladmandic/pilogger` which is my simple console logger (not required by `face-api` itself, i just use it in demos)
- `@tensorflow/tfjs-node` which is the actual tfjs engine
- `@vladmandic/face-api` which is the actual library

And `demo/node-canvas.js` adds one more on top of that to be able to write jpeg image outputs and not just text results
- `canvas`

**Comment by vladmandic** [comment]
@jayprout any updates?

**Comment by jayprout** [comment]
Hi vladmandic

My apologies for not responding sooner. The machine that I was trying the demo code on is at my work and we've shut down for the Christmas period. 

So I've tried to install it on another PC that I have here at home, and it's a PC that is only used for rendering 3D models, so it's a 'clean' machine insofar as it's a fresh Windows 10 64bit Operating System and until now the only software that's been installed on it has been the 3D model software and Google Chrome. 

So I've done the following and made notes as I went along:

downloaded and installed node v16.13.1x64 and I ticked the check box to auto install all necessary tools like Chocolatey etc.

After five minutes or so I get the following confirmation message:

Upgraded:
 - python v3.10.1
 - python3 v3.10.1
 - visualstudio2019-workload-vctools v1.0.1
 - chocolatey-visualstudio.extension v1.10.0
 - visualstudio2019buildtools v16.11.8.0
 - visualstudio-installer v2.0.2

As prompted I reboot the PC to finish the Node.js installation and its dependencies

I then downloaded and unzipped face-api-master.zip from https://github.com/vladmandic/face-api.git

I opened a command window and cd to face-api-master\demo

I ran the command:
npm init y

and then install the following modules as per your suggestion:
npm install @vladmandic/pilogger --save
npm install @tensorflow/tfjs-node --save
npm install @vladmandic/face-api --save

Everything seems to have installed ok and my package.json file inside face-api-master/demo reads like so
{
  "name": "demo",
  "version": "1.0.0",
  "description": "",
  "main": "node.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@tensorflow/tfjs-node": "^3.12.0",
    "@vladmandic/face-api": "^1.6.1",
    "@vladmandic/pilogger": "^0.4.2"
  }
}

I then run the command:
node node.js

and get the following error message

2021-12-26 17:07:16.994672: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '@tensorflow/tfjs-node'
Require stack:
- C:\iD6\face-api-master\dist\face-api.node.js
- C:\iD6\face-api-master\demo\node.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at dist/tfjs.esm.js (C:\iD6\face-api-master\dist\face-api.node.js:56:38)
    at __require (C:\iD6\face-api-master\dist\face-api.node.js:15:44)
    at Object.<anonymous> (C:\iD6\face-api-master\dist\face-api.node.js:184:23)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\iD6\\face-api-master\\dist\\face-api.node.js',
    'C:\\iD6\\face-api-master\\demo\\node.js'
  ]
}

And that's the current state of play. What am I doing wrong please?

**Comment by vladmandic** [comment]
it doesn't look like you're doing anything wrong, but the error doesn't make any sense  

> 2021-12-26 17:07:16.994672: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations

thats not an error, thats a silly info mesage that tensorflow prints when its loading. there are ways to remove it, but its not relevant right now.

>Error: Cannot find module '@tensorflow/tfjs-node'

now this is the error, but it makes no sense since clearly tensorflow DID get loaded otherwise it wouldn't print the previous message.

i'll run some tests and circle back in few days - like you said, its a holiday season :)

**Comment by jayprout** [comment]
Thank you Vladmandic for looking into this, enjoy the rest of the holiday season :)

Jason Prout
Create Technician
Create Team
Clear Channel UK

Sent on the go so please excuse brevity and spelling
________________________________

**Comment by sidequestlegend** [comment]
To me it looks like initialising the node.js demos from within the deo folder is the culprit here. The node.js demo references the dist files in the same repo, but those files are outside the demo folder and as such the above dependency on `@tensorflow/tfjs-node` breaks in that context. 
https://github.com/vladmandic/face-api/blob/master/demo/node.js#L10

It also seems clear the node example is intended to run from the root folder since the default sample image location references a subfolder in the current folder called demo:
https://github.com/vladmandic/face-api/blob/master/demo/node.js#L13

And the start script is executed in the root working directory too - 
https://github.com/vladmandic/face-api/blob/master/package.json#L24

TL;DR I think this is working correctly, when run from the root folder everything works as expected.

**Comment by vladmandic** [comment]
ah, i need to update comments in the demos - the imports such as 
```js
const faceapi = require('../dist/face-api.node.js');
```
are only meant to be used in dev environment when actually modifying and building `face-api`.

when using `face-api` as a module, it should simply be:
```js
const faceapi = require('@vladmandic/face-api');
```

of course, path to images is up to user (demos are indeed structured to load samples relative to project root folder).

anyhow, glad to figured it out...

**Comment by jayprout** [comment]
Happy New Year Vladmandic :)

I feel like such an idiot. Yes doing an npm init in the demo folder was where I was going wrong. Thank you so much for taking the time to show me the way, it's very much appreciated :)

**Comment by steamlocoSY** [comment]
Hi Vladmandic
I could not install @vlad/pilogger, is there any remedy for that module? (I am a freshman to Node.js :(
![IMG_20230802_181158](https://github.com/vladmandic/face-api/assets/126448953/87e1efc9-ede4-49d4-9826-315d6d9dd97b)

**Comment by steamlocoSY** [comment]
Hi Vladmandic
I could not install @vlad/pilogger, is there any remedy for that module? (I am a freshman to Node.js :(
![IMG_20230802_181158_edit_515141003068487](https://github.com/vladmandic/face-api/assets/126448953/b1321082-9bcc-4d0a-9d6f-6a11f02e2692)


Best regards
william
